### PR TITLE
Add pipeline overridable constants validation tests

### PR DESCRIPTION
**Pull request by shrekshao**
Issue:  None

[tint:1660](https://crbug.com/tint/1660): `compute_pipeline.spec.ts: overrides,workgroup_size,limits,workgroup_storage_size` is known to fail in Chrome.

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [ ] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [x] Tests are properly located in the test tree.
- [x] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [x] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [x] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/api/validation/compute_pipeline.spec.ts [hunk 1]
```diff
@@ -259,3 +259,254 @@ Tests calling createComputePipeline(Async) validation for compute workgroup_size
...
+      },
+    };
+
+    t.doCreateComputePipelineTest(isAsync, _success, descriptor);
+  });
+
+const kOverridesWorkgroupSizeShaderCode = `
+override x: u32 = 1u;
+override y: u32 = 1u;
+override z: u32 = 1u;
```
Comment (kainino0x) [inline src/webgpu/api/validation/compute_pipeline.spec.ts:374]: does workgroup_size also allow i32? if so, add a bit of coverage by changing one of these to i32

File: src/webgpu/api/validation/compute_pipeline.spec.ts [hunk 2]
```diff
@@ -259,3 +259,254 @@ Tests calling createComputePipeline(Async) validation for compute workgroup_size
...
+`
+  )
+  .params(u =>
+    u //
+      .combine('isAsync', [true, false])
+      .combineWithParams([
+        { constants: {}, _success: true },
+        { constants: { x: 0, y: 0, z: 0 }, _success: false },
+        { constants: { x: 16, y: 1, z: 1 }, _success: true },
+      ] as { constants: Record<string, GPUPipelineConstantValue>; _success: boolean }[])
```
Comment (kainino0x) [inline src/webgpu/api/validation/compute_pipeline.spec.ts:393]: nit: same

File: src/webgpu/api/validation/compute_pipeline.spec.ts [hunk 3]
```diff
@@ -259,3 +259,254 @@ Tests calling createComputePipeline(Async) validation for compute workgroup_size
...
+    `
+Tests calling createComputePipeline(Async) validation for overridable constants used for workgroup size.
+`
+  )
+  .params(u =>
+    u //
+      .combine('isAsync', [true, false])
+      .combineWithParams([
+        { constants: {}, _success: true },
+        { constants: { x: 0, y: 0, z: 0 }, _success: false },
```
Comment (kainino0x) [inline src/webgpu/api/validation/compute_pipeline.spec.ts:391]: would 0,1,1 or 1,0,1 or 1,0,0 also be invalid? perhaps add a few more cases

File: src/webgpu/api/validation/render_pipeline/overrides.spec.ts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        { vertexConstants: {}, _success: true },
+        { vertexConstants: { x: 1, y: 1 }, _success: true },
+        { vertexConstants: { x: 1, y: 1, 1: 1, 1000: 1 }, _success: true },
+        { vertexConstants: { xxx: 1 }, _success: false },
+        { vertexConstants: { 1: 1 }, _success: true },
+        { vertexConstants: { 2: 1 }, _success: false },
+        { vertexConstants: { z: 1 }, _success: false }, // pipeline constant id is specified for z
+        { vertexConstants: { w: 1 }, _success: false }, // pipeline constant id is specified for w
+        { vertexConstants: { 1: 1, z: 1 }, _success: false }, // pipeline constant id is specified for z
+      ] as { vertexConstants: Record<string, GPUPipelineConstantValue>; _success: boolean }[])
```
Comment (kainino0x) [inline src/webgpu/api/validation/render_pipeline/overrides.spec.ts:30]: nit: same

File: src/webgpu/api/validation/compute_pipeline.spec.ts [hunk 5]
```diff
@@ -259,3 +259,254 @@ Tests calling createComputePipeline(Async) validation for compute workgroup_size
...
+  )
+  .params(u =>
+    u //
+      .combine('isAsync', [true, false])
+      .combineWithParams([
+        { constants: {}, _success: false },
+        { constants: { c0: 0, c2: 0, c8: 0 }, _success: false }, // c5 is missing
+        { constants: { c0: 0, c2: 0, c5: 0, c8: 0 }, _success: true },
+        { constants: { c0: 0, c2: 0, c5: 0, c8: 0, c1: 0 }, _success: true },
+      ] as { constants: Record<string, GPUPipelineConstantValue>; _success: boolean }[])
```
Comment (kainino0x) [inline src/webgpu/api/validation/compute_pipeline.spec.ts:327]: nit: same

File: src/webgpu/api/validation/compute_pipeline.spec.ts [hunk 6]
```diff
@@ -259,3 +259,254 @@ Tests calling createComputePipeline(Async) validation for compute workgroup_size
...
+        { constants: { c0: 0 }, _success: true },
+        { constants: { c0: 0, c1: 1 }, _success: true },
+        { constants: { c9: 0 }, _success: false },
+        { constants: { 1: 0 }, _success: true },
+        { constants: { c3: 0 }, _success: false }, // pipeline constant id is specified for c3
+        { constants: { 2: 0 }, _success: false },
+        { constants: { 1000: 0 }, _success: true },
+        { constants: { 9999: 0 }, _success: false },
+        { constants: { 1000: 0, c2: 0 }, _success: false },
+      ] as { constants: Record<string, GPUPipelineConstantValue>; _success: boolean }[])
```
Comment (kainino0x) [inline src/webgpu/api/validation/compute_pipeline.spec.ts:283]: nit: would simply `as const` work instead? (I think it should)

File: src/webgpu/api/validation/shader_module/overrides.spec.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+@id(1234) override c1: u32;
+
+@compute @workgroup_size(1) fn main() {
+  // make sure the overridable constants are not optimized out
+  _ = c0;
+  _ = c1;
+}
+          `,
+      });
+    }, true);
```
Comment (kainino0x) [inline src/webgpu/api/validation/shader_module/overrides.spec.ts:45]: do we have a similar test for two overrides with the same name (like, `c0` and `c0`)?

**Review by kainino0x — APPROVED** [review]
LGTM with a few suggestions

File: src/webgpu/api/validation/compute_pipeline.spec.ts (hunk 1 above)
Comment (shrekshao) [reply src/webgpu/api/validation/compute_pipeline.spec.ts:374]: Done. They need to either all being u32 or all being i32. Test with 2 shaders.

File: src/webgpu/api/validation/compute_pipeline.spec.ts (hunk 6 above)
Comment (shrekshao) [reply src/webgpu/api/validation/compute_pipeline.spec.ts:283]: Turns out no. This is the error log from vscode intellisense. `as const` seems to infer different types for each constants (dictionary with different keys)

```
Argument of type '{ layout: "auto"; compute: { module: GPUShaderModule; entryPoint: string; constants: { readonly x?: undefined; readonly y?: undefined; readonly z?: undefined; } | { readonly x: 0; readonly y: 0; readonly z: 0; } | { ...; } | { ...; } | { ...; }; }; }' is not assignable to parameter of type 'GPUComputePipelineDescriptor'.
```

**Pull request merged**